benchexec/resources.py

```diff
@@ -210,118 +218,738 @@ def _get_cpu_cores_per_run0(
+                ].copy()
+                for core in core_with_siblings:
+                    if len(cores) < coreLimit:
+                        cores.append(core)  # add core&siblings to results
+                    else:
+                        blocked_cores.append(
+                            core
+                        )  # add superfluous cores to blocked_cores
+                    core_clean_up(core, allCpus, hierarchy_levels)
+
+            while sub_unit_cores:
+                core_clean_up(sub_unit_cores[0], allCpus, hierarchy_levels)
+                # active_cores remove(sub_unit_cores[0])
+                # sub_unit_cores remove(sub_unit_cores[0])
+
+            # if coreLimit reached: append core to result, delete remaining cores from active_cores
+            if len(cores) == coreLimit:
+                result.append(cores)
+
+    # cleanup: while-loop stops before running through all units: while some active_cores-lists
+    # & sub_unit_cores-lists are empty, other stay half-full or full
+    logging.debug(f"Core allocation:{result}")
+    return result
+
+
+def check_symmetric_num_of_values(hierarchy_level: HierarchyLevel) -> bool:
+    """returns True if the number of values in the lists of the key-value pairs
+    is equal throughout the dict"""
+    return not check_asymmetric_num_of_values(hierarchy_level)
+
+
+def check_asymmetric_num_of_values(hierarchy_level: HierarchyLevel) -> bool:
+    """returns True if the number of values in the lists of the key-value pairs
+    is not equal throughout the dict"""
+    is_asymmetric = False
+    cores_per_unit = len(next(iter(hierarchy_level.values())))
+    if any(len(cores) != cores_per_unit for cores in hierarchy_level.values()):
+        is_asymmetric = True
+    return is_asymmetric
+
+
+def core_clean_up(
+    core: int,
+    allCpus: Dict[int, VirtualCore],
+    hierarchy_levels: List[HierarchyLevel],
+) -> None:
+    current_core_regions = allCpus[core].memory_regions
+    for mem_index in range(len(current_core_regions)):
+        region = current_core_regions[mem_index]
+        hierarchy_levels[mem_index][region].remove(core)
+        if len(hierarchy_levels[mem_index][region]) == 0:
+            hierarchy_levels[mem_index].pop(region)
+
+
+# return list of available CPU cores
+def get_cpu_list(my_cgroups, coreSet: Optional[List] = None) -> List[int]:
+    # read list of available CPU cores
+    allCpus = my_cgroups.read_allowed_cpus()
+
+    # Filter CPU cores according to the list of identifiers provided by a user
+    if coreSet:
+        invalid_cores = sorted(set(coreSet).difference(set(allCpus)))
+        if len(invalid_cores) > 0:
+            raise ValueError(
+                "The following provided CPU cores are not available: "
+                + ", ".join(map(str, invalid_cores))
+            )
+        allCpus_list = [core for core in allCpus if core in coreSet]
+    allCpus_list = frequency_filter(allCpus, 0.05)
```

Comment:
Extended the docstring of get_cpu_list:     
A filter is applied to make sure, all cores used for the benchmark run at the same clock speed (allowing a deviation of 0.05 (5%) from the highest frequency)

Additionally, the docstring of of frequency_filter: 

Filters the list of all available CPU cores so that only the fastest cores are used for the benchmark run.
Only cores with a maximal frequency within the distance of the defined threshold from the maximal frequency of the fastest core are added to the filtered_allCpus_list and returned for further use. (max_frequency of core) >= (1-threshold)*(max_frequency of fastest core).
All cores that are slower will not be used for the benchmark and displayed in a debug message.

@param allCpus_list: list of all cores available for the benchmark run
@param threshold: accepted difference (as percentage) in the maximal frequency of a core from the fastest core to still be used in the benchmark run
@return: filtered_allCpus_list with only the fastest cores

Reply:
See commit https://github.com/sosy-lab/benchexec/pull/892/commits/0689df517323f46784726a9fc5b3d74c0abcdad0

Reply:
Further improved in 4a6414232aec50678971a9e3e7867fc079cee7fe because the magic constant should be only in one place.